--- a/ipq806x/nss_gmac_ctrl.c
+++ b/ipq806x/nss_gmac_ctrl.c
@@ -1213,6 +1213,8 @@
 		return -ENOMEM;
 	}
 
+	SET_NETDEV_DEV(netdev, &pdev->dev);
+
 	gmacdev = netdev_priv(netdev);
 	memset((void *)gmacdev, 0, sizeof(struct nss_gmac_dev));
 
@@ -1439,8 +1441,6 @@
 		netdev_dbg(netdev, "%s MII_PHYSID2 - 0x%04x\n", netdev->name,
 		      nss_gmac_mii_rd_reg(gmacdev, gmacdev->phy_base, MII_PHYSID2));
 	} else if (gmacdev->phy_base != NSS_GMAC_NO_MDIO_PHY) {
-		SET_NETDEV_DEV(netdev, gmacdev->miibus->parent);
-
 		/*
 		 * Issue a phy_attach for the interface connected to a switch
 		 */
