https://github.com/torvalds/linux/commit/a44d9e72100f7044ac46e4e6dc475f5b4097830f
--- a/openvpn/plugins/nss_ovpn_sk.c
+++ b/openvpn/plugins/nss_ovpn_sk.c
@@ -773,7 +773,6 @@ static const struct proto_ops nss_ovpn_sk_proto_ops = {
 	.ioctl = nss_ovpn_sk_ioctl,
 	.listen = sock_no_listen,
 	.shutdown = sock_no_shutdown,
-	.getsockopt = sock_no_getsockopt,
 	.mmap = sock_no_mmap,
 	.sendpage = sock_no_sendpage,
 	.sendmsg = nss_ovpn_sk_sendmsg,
@@ -781,7 +780,6 @@ static const struct proto_ops nss_ovpn_sk_proto_ops = {
 	.poll = datagram_poll,
 	.bind = sock_no_bind,
 	.release = nss_ovpn_sk_release,
-	.setsockopt = sock_no_setsockopt,
 	.accept = sock_no_accept,
 };
 
