--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -1335,8 +1335,10 @@ struct ath10k {
 
 static inline bool ath10k_peer_stats_enabled(struct ath10k *ar)
 {
-	if (test_bit(ATH10K_FLAG_PEER_STATS, &ar->dev_flags) &&
-	    test_bit(WMI_SERVICE_PEER_STATS, ar->wmi.svc_map))
+	if (test_bit(ATH10K_FLAG_PEER_STATS, &ar->dev_flags))
+		return true;
+
+	if (test_bit(WMI_SERVICE_PEER_STATS, ar->wmi.svc_map))
 		return true;
 
 	return false;
