--- a/drivers/net/wireless/ath/ath10k/debug.c
+++ b/drivers/net/wireless/ath/ath10k/debug.c
@@ -94,12 +94,14 @@ void ath10k_debug_print_board_info(struc
 
 void ath10k_debug_print_boot_info(struct ath10k *ar)
 {
-	ath10k_info(ar, "htt-ver %d.%d wmi-op %d htt-op %d cal %s max-sta %d raw %d hwcrypto %d\n",
+	ath10k_info(ar,
+		    "htt-ver %d.%d wmi-op %d htt-op %d cal %s max-peer %d max-sta %d raw %d hwcrypto %d\n",
 		    ar->htt.target_version_major,
 		    ar->htt.target_version_minor,
 		    ar->normal_mode_fw.fw_file.wmi_op_version,
 		    ar->normal_mode_fw.fw_file.htt_op_version,
 		    ath10k_cal_mode_str(ar->cal_mode),
+		    ar->max_num_peers,
 		    ar->max_num_stations,
 		    test_bit(ATH10K_FLAG_RAW_MODE, &ar->dev_flags),
 		    !test_bit(ATH10K_FLAG_HW_CRYPTO_DISABLED, &ar->dev_flags));
--- a/drivers/net/wireless/ath/ath10k/smart_ant.c
+++ b/drivers/net/wireless/ath/ath10k/smart_ant.c
@@ -1015,11 +1015,11 @@ static u8 smart_ant_proc_train_stats(str
 		if (ar->smart_ant_info.debug_level >=
 		    ATH10K_SMART_ANT_DBG_LVL_TOP_DECISION) {
 			ath10k_dbg(ar, ATH10K_DBG_SMART_ANT,
-				   "Training completed:antenna selected %d PER %d RSSI %d %d %d rate %x\n",
+				   "Training completed:antenna selected %d PER %d RSSI %d %d %d rate %x, max peers %d, max sta %d\n",
 				   tstats->ant_map[0],
 				   tstats->per,
 				   tstats->rssi[0][0], tstats->rssi[1][0],
-				   tstats->rssi[2][0], tdata->rate_code);
+				   tstats->rssi[2][0], tdata->rate_code, ar->max_num_peers, ar->max_num_stations);
 		}
 		/* force config tx/rx antenna after SA decision is made */
 		status |= ATH10K_SMART_ANT_ACT_TX_CFG;
