--- a/arch/arm/boot/dts/qcom/qcom-ipq8064.dtsi
+++ b/arch/arm/boot/dts/qcom/qcom-ipq8064.dtsi
@@ -406,6 +406,11 @@
 
 	memory {
 		device_type = "memory";
+		/* coreboot value reg = <0x41500000 0x18a00000 0x5a000000 0x26000000>; */
+		/* get coreboot value - cat /sys/firmware/fdt > /tmp/test.dtb */
+		/* with HACK force ZRELADDR on arch-qcom use <0x40000000 0x38000000> */
+		/* with openwrt commit b2bb4b0f0a4ecd4c3ad58ea42dde14dce87788fc use 0x42000000 */
+		linux,usable-memory = <0x42000000 0x38000000>;
 		reg = <0x0 0x0>;
 	};
 
