--- a/ecm_classifier_mscs.c
+++ b/ecm_classifier_mscs.c
@@ -429,9 +429,12 @@ static void ecm_classifier_mscs_process(struct ecm_classifier_instance *aci, ecm_tracker_sender_type_t sender,
 	struct ecm_db_connection_instance *ci = NULL;
 	struct ecm_front_end_connection_instance *feci;
 	ecm_front_end_acceleration_mode_t accel_mode;
+	int protocol;
 	uint32_t became_relevant = 0;
 	ecm_classifier_mscs_process_callback_t cb = NULL;
+#ifdef ECM_CLASSIFIER_MSCS_SCS_ENABLE
 	bool scs_result = false;
+#endif
 	ecm_classifier_mscs_result_t result = 0;
 	uint8_t smac[ETH_ALEN];
 	uint8_t dmac[ETH_ALEN];
