--- a/src/init/ssdk_init.c
+++ b/src/init/ssdk_init.c
@@ -293,7 +293,11 @@ qca_ar8327_phy_fixup(struct qca_phy_priv *priv, int phy)
 	case 2:
 		priv->phy_mmd_write(priv->device_id, phy, 0x7, 0x3c);
 		priv->phy_mmd_write(priv->device_id, phy, 0x4007, 0x0);
+#if __has_attribute(__fallthrough__)
+		__attribute__((__fallthrough__));
+#else
 		/* fallthrough */
+#endif
 	case 4:
 		if(priv->version == QCA_VER_AR8327)
 		{
--- a/src/init/ssdk_plat.c
+++ b/src/init/ssdk_plat.c
@@ -989,6 +989,7 @@ static ssize_t ssdk_phy_write_reg_set(struct device *dev,
 	char *this_opt;
 	char *options = phy_buf;
 	unsigned int phy_addr, reg_addr, reg_value;
+	int ret;
 
 	if (count >= sizeof(phy_buf))
 		return 0;
@@ -999,7 +1000,10 @@ static ssize_t ssdk_phy_write_reg_set(struct device *dev,
 	if (!this_opt)
 		goto fail;
 
-	kstrtouint(this_opt, 0, &phy_addr);
+	ret = kstrtouint(this_opt, 0, &phy_addr);
+	if (ret)
+		goto fail;
+
 	if ((options - phy_buf) >= (count - 1))
 		goto fail;
 
@@ -1007,7 +1011,10 @@ static ssize_t ssdk_phy_write_reg_set(struct device *dev,
 	if (!this_opt)
 		goto fail;
 
-	kstrtouint(this_opt, 0, &reg_addr);
+	ret = kstrtouint(this_opt, 0, &reg_addr);
+	if (ret)
+		goto fail;
+
 	if ((options - phy_buf) >= (count - 1))
 		goto fail;
 
@@ -1015,7 +1022,9 @@ static ssize_t ssdk_phy_write_reg_set(struct device *dev,
 	if (!this_opt)
 		goto fail;
 
-	kstrtouint(this_opt, 0, &reg_value);
+	ret = kstrtouint(this_opt, 0, &reg_value);
+	if (ret)
+		goto fail;
 
 	qca_ar8327_phy_write(0, phy_addr, reg_addr, reg_value);
 
@@ -1044,6 +1053,7 @@ static ssize_t ssdk_phy_read_reg_set(struct device *dev,
 	char *this_opt;
 	char *options = phy_buf;
 	unsigned int phy_addr, reg_addr;
+	int ret;
 
 	if (count >= sizeof(phy_buf))
 		return 0;
@@ -1054,7 +1064,10 @@ static ssize_t ssdk_phy_read_reg_set(struct device *dev,
 	if (!this_opt)
 		goto fail;
 
-	kstrtouint(this_opt, 0, &phy_addr);
+	ret = kstrtouint(this_opt, 0, &phy_addr);
+	if (ret)
+		goto fail;
+
 	if ((options - phy_buf) >= (count - 1))
 		goto fail;
 
@@ -1062,7 +1075,9 @@ static ssize_t ssdk_phy_read_reg_set(struct device *dev,
 	if (!this_opt)
 		goto fail;
 
-	kstrtouint(this_opt, 0, &reg_addr);
+	ret = kstrtouint(this_opt, 0, &reg_addr);
+	if (ret)
+		goto fail;
 
 	qca_ar8327_phy_read(0, phy_addr, reg_addr, &phy_reg_val);
 
